### PR TITLE
feat: Use AdvancedSearch DocumentBuilder from UpdateResourceInIndex if available

### PR DESCRIPTION
**Pull request by hectoras**
**Associated Jira issue:** [PISA25-188](https://oat-sa.atlassian.net/browse/PISA25-188)

Related PRs:
- https://github.com/oat-sa/extension-tao-advanced-search/pull/86

### CODE REVIEW

File: models/classes/search/index/DocumentBuilder/IndexDocumentBuilder.php [hunk 1]
```diff
@@ -74,32 +74,43 @@ public function createDocumentFromResource(Resource $resource): IndexDocument
...
+        $resource = new Resource($resourceData['id']);
+
+        if (isset($resourceData['indexProperties'])) {
+            $indexProperties = $resourceData['indexProperties'];
+        } else {
+            $tokenizationInfo = $this->getTokenizedResourceBody($resource);
+            $indexProperties = $tokenizationInfo['indexProperties'];
+        }
+
+        $this->logDebug('createDocumentFromArray');
```
Comment (hectoras) [inline models/classes/search/index/DocumentBuilder/IndexDocumentBuilder.php:106]: TODO: Remove this

File: models/classes/resources/ResourceWatcher.php [hunk 2]
```diff
@@ -75,22 +75,25 @@ public function catchUpdatedResourceEvent(ResourceUpdated $event): void
...
+            $this->getLogger()->debug(
+                'triggering index update on resourceUpdated event'
+            );
 
             $property = $this->getProperty(TaoOntology::PROPERTY_UPDATED_AT);
             $this->updatedAtCache[$resource->getUri()] = $now;
             $resource->editPropertyValues($property, $now);
+
+            $taskMessage = __('Adding/updating search index for updated resource');
+            $this->createResourceIndexingTask($resource, $taskMessage);
```
Comment (gabrielfs7) [inline models/classes/resources/ResourceWatcher.php:96]: wait a minute... We are interested in catching the update events for ItemUpdated event + TestUpdated event, but now we are introducing a new catch to `ResourceUpdated` that did not exist before?

So why we did not just handled the `TestUpdatedEvent` and `ItemUpdated` event in taoAdvancedSearch? Why do we need this change here?

File: models/classes/search/tasks/UpdateResourceInIndex.php [hunk 3]
```diff
@@ -57,7 +66,68 @@ public function __invoke($params): Report
...
+            $document->getDynamicProperties(),
+            $document->getAccessProperties()
+        );
+    }
+
+    private function getBody(
+        core_kernel_classes_Resource $resource,
+        IndexDocument $document
+    ): array {
+        if ($this->hasReferencesService()) {
```
Comment (gabrielfs7) [inline models/classes/search/tasks/UpdateResourceInIndex.php:119]: Can we put some TODO here to remove this later, please?

File: models/classes/search/tasks/UpdateResourceInIndex.php [hunk 4]
```diff
@@ -89,38 +159,41 @@ public function __invoke($params): Report
...
+    }
 
-        return $documents;
+    private function getIndexService(): IndexService
+    {
+        return $this->getService(IndexService::SERVICE_ID);
     }
 
-    private function getIndexDocumentBuilder(): IndexDocumentBuilder
+    private function getReferencesService(): ResourceReferencesService
```
Comment (gabrielfs7) [inline models/classes/search/tasks/UpdateResourceInIndex.php:180]: Same here https://github.com/oat-sa/tao-core/pull/3673/files?diff=split&w=1#r1030688521

File: models/classes/search/tasks/UpdateResourceInIndex.php [hunk 5]
```diff
@@ -89,38 +159,41 @@ public function __invoke($params): Report
...
     {
-        $documentBuilder = $this->getIndexService()->getDocumentBuilder();
-        $documentBuilder->setServiceLocator($this->getServiceLocator());
+        return $this->getService(ResourceReferencesService::class);
+    }
 
-        return $documentBuilder;
+    private function hasReferencesService(): bool
```
Comment (gabrielfs7) [inline models/classes/search/tasks/UpdateResourceInIndex.php:185]: and here https://github.com/oat-sa/tao-core/pull/3673/files?diff=split&w=1#r1030688521

**Comment by hectoras** [comment]
Rebased with `develop`.

**Comment by hectoras** [comment]
Code coverage is again above the minimum allowed threshold, and new code conforms to PSR-12.

File: models/classes/search/tasks/UpdateResourceInIndex.php [hunk 6]
```diff
@@ -57,7 +63,36 @@ public function __invoke($params): Report
...
+
+        foreach ($resourceUris as $resourceUri) {
+            $this->resourceUris[] = $resourceUri;
+
+            $documents[] = $this->getDocumentBuilder()->createDocumentFromResource(
+                $this->getResource($resourceUri)
+            );
+        }
+
+        return $documents;
```
Comment (gabrielfs7) [inline models/classes/search/tasks/UpdateResourceInIndex.php:91]: I know I wrote this code, but I should have optimized it :) 

```suggestion
        $documentBuilder = $this->getDocumentBuilder();
        $documents = [];

        foreach ($resourceUris as $resourceUri) {
            $this->resourceUris[] = $resourceUri;

            $documents[] = $documentBuilder->createDocumentFromResource($this->getResource($resourceUri));
        }

        return $documents;
```

**Review by gabrielfs7 — APPROVED** [review]

**Comment by hectoras** [comment]
Released as [v50.29.0](https://github.com/oat-sa/tao-core/releases/tag/v50.29.0)

**Pull request merged**